### ISSUE TITLE
cannot feed data with correct raw samples

### ISSUE THREAD

**Issue by matjazPrijatelj** [samirkumardas/jmuxer#41]
Hi

I want to feed data from web socket - 1 nal unit at a time from ws proxy server which forwards rtp stream.

//index.html
`    this.wsSocket = new WebSocket("ws://localhost:9999");
    this.wsSocket.binaryType = "arraybuffer";
    this.wsSocket.addEventListener("message", this.feed);

    feed(event) {
      let data = new Uint8Array(event.data);
      jmuxer.feed({
        video: data,
      });
    },`

I'm sending this sequence which works properly in broadway.js decoder.

// example payloads from server
<Buffer 00 00 01 21 9a 10 3c 00 70 87 c0>
<Buffer 00 00 01 21 9a 12 3c 00 70 87 c0>
<Buffer 00 00 00 01 27 42 80 28 95 a0 14 01 6e c0 40 78 91 35>
<Buffer 00 00 00 01 28 ce 02 5c 80>
<Buffer 00 00 01 25 88 80 00 01 ea b4 ff ff fc 3d 14 00 0d ff ff fe 4f ff fc 9f ff f9 3f ff f2 7f ff e4 ff ff c9 ff ff 93 ff ff 27 ff fe 4f ff fc 9f ff f9 3f ... 10315 more bytes>
<Buffer 00 00 01 21 9a 02 3c 00 70 87 c0>
<Buffer 00 00 01 21 9a 04 3c 00 70 87 c0>

I don't have any error message just no image. After quick debug session I only found that no SPS and PPS frames are picked. Are they essential to get the stuff working or is fps option sufficiant?

I've read your answer regarding reading sps and pps, which isn't read in my example when putting breakpoint on parseNAL().
https://github.com/samirkumardas/jmuxer/issues/36#issuecomment-601575064

correct type is picked if reading ntype on nalu.js constructor.

This looks my converted buffer which is feeding:
Uint8Array(17) [0, 0, 1, 39, 66, 128, 40, 149, 160, 20, 1, 110, 192, 64, 120, 145, 53] - SPS
Uint8Array(8) [0, 0, 1, 40, 206, 2, 92, 128] - PPS

**Comment by samirkumardas** [comment]
You must have SPS and PPS to work. 
Broadway decoder renders the frame image and plays it using WebGL or canvas whereas JMuxer uses MSE extensions to the play video. In layman terms, SPS and PPS contain some meta-information e.g. frame dimension, the codec, etc. that are essential when you encapsulate media data into the MP4 container.

**Comment by samirkumardas** [comment]
As it does not show the error message and SPS and PPS are available, have you clicked the play button to play?  It may not play automatically in some browser depending on the browser privacy/security setting

**Comment by matjazPrijatelj** [comment]
Hi thank you for reply.

I tried to click play - no go.

is a class constructor ok, for a video only feed:

    this.jmuxer = new JMuxer({
      node: "player",
      mode: "video",
      flushingTime: 1000,
      fps: 30,
      debug: true,
    });

I've put a breakpoint on parseSDP() but I never see any unit with SPS,PPS

this is feeding into jmuxer feed: 

`[[Uint8Array]]: Uint8Array(18) [0, 0, 0, 1, 39, 66, 128, 40, 149, 160, 20, 1, 110, 192, 64, 
[[Uint8Array]]: Uint8Array(9) [0, 0, 0, 1, 40, 206, 2, 92, 128]
[[Uint8Array]]: Uint8Array(24160) [0, 0, 1, 37, 136, 128, 0, 21, 48, 255, 255, 252, 45, 20, 0, 63, 92, 112, 199, 31, 255, 254, 79, 255, 252, 35, 239, 255, …]`

**Comment by samirkumardas** [comment]
NAL types 1 and 5 are IDR and NDR respectively. So the commented line should have been true.

It seems you get fewer frames whereas your fps are 30 which seems to be high. So lagging is expected.

**Comment by matjazPrijatelj** [comment]
Yes I am aware that that should be true but NAL's with SPS and PPS aren't picked before that. Where should I check why they aren't processed before playback starts.

**Comment by matjazPrijatelj** [comment]
I've added: 

`  getVideoFrames(nalus, duration) {
    let nalu,
      units = [],
      samples = [],
      naluObj,
      sampleDuration,
      adjustDuration = 0,
      numberOfFrames = [];

    for (nalu of nalus) {
      naluObj = new NALU(nalu);
      // console.log(naluObj); // proper

      units.push(naluObj);

      if (naluObj.type() === NALU.SPS) {
        this.remuxController.tracks.video.h264.parseNAL(naluObj);
      } else if (naluObj.type() === NALU.PPS) {
        this.remuxController.tracks.video.h264.parseNAL(naluObj);
      }

      if (naluObj.type() === NALU.IDR || naluObj.type() === NALU.NDR) {
        samples.push({ units });
        units = [];
        if (this.options.clearBuffer) {
          if (naluObj.type() === NALU.IDR) {
            numberOfFrames.push(this.frameCounter);
          }
          this.frameCounter++;
        }
      }
    }`

Since otherwise no SPS or PPS was triggered in parseNAL() method.

I've decreased flushing time to 1 to reduce delay. Now it seems to work.

**Comment by samirkumardas** [comment]
1. Glad to hear you have found a workaround. Though I am not sure why jmuxer failed to detect sps/pps.  According to your code, it should have.

2.  Small value `flushingTime` can reduce latency but it will add fewer frames to the media buffer at a time.  As you have fewer frames, you can keep it a little value. 

3. `clearBuffer` is essential if you are playing live streaming. It must be set to true, otherwise, it may swallow memory gradually.  However, there is an issue pertains to the browser inactive tab.  You can check this issue https://github.com/samirkumardas/jmuxer/issues/40
Jmuxer don't have a solution right now, I have to work on this. You can follow that post if it helps.

4. As you are using a fixed FPS, it may create a latency gradually unless you provide a steady data to the jumxer. Moreover, network latency or interruption could introduce additional latency as jumxer does not filter out the stale frame itself. Therefore, you have to implement your own logic to remove the stale frames before feeding to jmuxer.  It is out of the scope of jmuxer.  In a standard case, you must provide data as a chunk and its duration (chunk duration) instead of using a fixed fps.  Unfortunately, I don't have any example regarding this on Github.

**Comment by matjazPrijatelj** [comment]
Thank you I'll deffinitely dig more into this.
